File: CGConvertor/MediaInspector.swift
import Foundation

/// Inspecție/Metadata profundă + thumbnail cu LUT static (Faza 2). Rulează
/// `ffprobe`/`ffmpeg` — ACELAȘI binar deja folosit de `MotorFFmpeg` pentru
/// conversie, fără nicio dependință nouă. Analiza rulează asincron, în
/// fundal, imediat după ce un fișier intră în coadă (vezi
/// `ConvertorViewModel.adaugaFisiere`) — nu blochează UI-ul niciodată.
struct MediaMetadata: Equatable {
    var durataSecunde: Double?
    var codecVideo: String?
    var latime: Int?
    var inaltime: Int?
    var frameRate: String?      // ex. "23.976", derivat din r_frame_rate (fracție)
    var bitrateVideo: Int?      // bps
    var pixFmt: String?
    var colorSpace: String?
    var colorTransfer: String?
    var colorPrimaries: String?
    var codecAudio: String?
    var sampleRateAudio: Int?
    var canaleAudio: Int?
    var bitrateAudio: Int?

    var rezolutieText: String? {
        guard let l = latime, let h = inaltime else { return nil }
        return "\(l)×\(h)"
    }
}

enum MediaInspector {
    /// Rulează `ffprobe -show_format -show_streams` și parsează JSON-ul
    /// brut manual (nu Codable pe structura ffprobe direct — câmpurile ei
    /// variază mult între containere/codecuri, un parsing tolerant e mai
    /// robust decât un model rigid care ar eșua la primul câmp lipsă).
    static func probe(url: URL) -> MediaMetadata? {
        guard let ffmpegPath = MotorFFmpeg.gasesteBinar() else { return nil }
        let ffprobePath = ffmpegPath.replacingOccurrences(of: "ffmpeg", with: "ffprobe")
        guard FileManager.default.fileExists(atPath: ffprobePath) else { return nil }

        let proces = Process()
        proces.executableURL = URL(fileURLWithPath: ffprobePath)
        proces.arguments = [
            "-v", "error", "-print_format", "json",
            "-show_format", "-show_streams",
            url.path
        ]
        let pipe = Pipe()
        proces.standardOutput = pipe
        proces.standardError = Pipe()

        do {
            try proces.run()
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            proces.waitUntilExit()
            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
            return parse(json)
        } catch {
            return nil
        }
    }

    private static func parse(_ json: [String: Any]) -> MediaMetadata {
        var meta = MediaMetadata()
        if let format = json["format"] as? [String: Any] {
            meta.durataSecunde = (format["duration"] as? String).flatMap(Double.init)
        }
        guard let streams = json["streams"] as? [[String: Any]] else { return meta }

        if let v = streams.first(where: { ($0["codec_type"] as? String) == "video" }) {
            meta.codecVideo = v["codec_name"] as? String
            meta.latime = v["width"] as? Int
            meta.inaltime = v["height"] as? Int
            meta.pixFmt = v["pix_fmt"] as? String
            meta.colorSpace = v["color_space"] as? String
            meta.colorTransfer = v["color_transfer"] as? String
            meta.colorPrimaries = v["color_primaries"] as? String
            if let br = v["bit_rate"] as? String { meta.bitrateVideo = Int(br) }
            if let rate = v["r_frame_rate"] as? String {
                let parts = rate.split(separator: "/")
                if parts.count == 2, let num = Double(parts[0]), let den = Double(parts[1]), den != 0 {
                    meta.frameRate = String(format: "%.3f", num / den)
                }
            }
        }
        if let a = streams.first(where: { ($0["codec_type"] as? String) == "audio" }) {
            meta.codecAudio = a["codec_name"] as? String
            meta.canaleAudio = a["channels"] as? Int
            if let sr = a["sample_rate"] as? String { meta.sampleRateAudio = Int(sr) }
            if let br = a["bit_rate"] as? String { meta.bitrateAudio = Int(br) }
        }
        return meta
    }

    /// Extrage un cadru static (~1s în clip, sau primul cadru dacă mai
    /// scurt) ca thumbnail JPEG, opțional cu un LUT `.cube` aplicat prin
    /// filtrul nativ `lut3d` al FFmpeg — NU un player real-time (acela
    /// rămâne un TODO separat, mult mai mare, vezi CLAUDE.md).
    static func genereazaThumbnail(url: URL, lutPath: String?, iesire: URL) -> Bool {
        guard let ffmpegPath = MotorFFmpeg.gasesteBinar() else { return false }
        var filtru = "scale=320:-2"
        if let lutPath, !lutPath.isEmpty {
            let caleEscapata = lutPath.replacingOccurrences(of: "'", with: "\\'")
            filtru += ",lut3d=file='\(caleEscapata)'"
        }
        let proces = Process()
        proces.executableURL = URL(fileURLWithPath: ffmpegPath)
        proces.arguments = [
            "-y", "-ss", "1", "-i", url.path,
            "-frames:v", "1", "-vf", filtru,
            iesire.path
        ]
        proces.standardOutput = Pipe()
        proces.standardError = Pipe()
        do {
            try proces.run()
            proces.waitUntilExit()
            return proces.terminationStatus == 0 && FileManager.default.fileExists(atPath: iesire.path)
        } catch {
            return false
        }
    }

    /// Folder de lucru pentru thumbnail-uri, curățat la fiecare lansare
    /// (nu sunt persistente — regenerate din surse la nevoie).
    static func folderThumbnailuri() -> URL {
        let base = FileManager.default.temporaryDirectory.appendingPathComponent("CGConvertorThumbs", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }
}

File: python/media_inspector.py
# media_inspector.py
"""Inspectie/Metadata profunda + thumbnail cu LUT static (Faza 2). Ruleaza
ffprobe/ffmpeg — ACELEASI binare deja folosite de converter.py pentru
conversie, fara nicio dependinta noua. Port 1:1 al `MediaInspector.swift`
(Mac) — vezi acolo pentru comentariile complete de design."""

import base64
import json
import os
import subprocess
import sys
import tempfile
import time

from converter import get_ffmpeg_path, get_ffprobe_path

# Vezi comentariul din converter.py — acelasi fix pentru "fereastra
# neagra care clipeste" pe Windows (aici, cel mai vizibil: un apel
# ffmpeg PER MISCARE de slider in previzualizarea interactiva).
_NO_WINDOW = subprocess.CREATE_NO_WINDOW if sys.platform == "win32" else 0


def probe(path):
    """Ruleaza ffprobe -show_format -show_streams si parseaza JSON-ul brut
    manual (nu un model rigid) — campurile difera mult intre containere/
    codecuri, un parsing tolerant esueaza mai gratios decat unul strict."""
    ffprobe_path = get_ffprobe_path()
    try:
        result = subprocess.run(
            [ffprobe_path, "-v", "error", "-print_format", "json",
             "-show_format", "-show_streams", path],
            capture_output=True, text=True, check=True, creationflags=_NO_WINDOW,
        )
        data = json.loads(result.stdout)
    except (subprocess.CalledProcessError, FileNotFoundError, json.JSONDecodeError):
        return None

    meta = {
        "duration": None, "video_codec": None, "width": None, "height": None,
        "frame_rate": None, "video_bitrate": None, "pix_fmt": None,
        "color_space": None, "color_transfer": None, "color_primaries": None,
        "audio_codec": None, "sample_rate": None, "channels": None, "audio_bitrate": None,
    }
    fmt = data.get("format", {})
    try:
        meta["duration"] = float(fmt.get("duration"))
    except (TypeError, ValueError):
        pass

    streams = data.get("streams", [])
    video = next((s for s in streams if s.get("codec_type") == "video"), None)
    if video:
        meta["video_codec"] = video.get("codec_name")
        meta["width"] = video.get("width")
        meta["height"] = video.get("height")
        meta["pix_fmt"] = video.get("pix_fmt")
        meta["color_space"] = video.get("color_space")
        meta["color_transfer"] = video.get("color_transfer")
        meta["color_primaries"] = video.get("color_primaries")
        try:
            meta["video_bitrate"] = int(video.get("bit_rate"))
        except (TypeError, ValueError):
            pass
        rate = video.get("r_frame_rate", "")
        if "/" in rate:
            num, _, den = rate.partition("/")
            try:
                num, den = float(num), float(den)
                if den:
                    meta["frame_rate"] = f"{num / den:.3f}"
            except ValueError:
                pass

    audio = next((s for s in streams if s.get("codec_type") == "audio"), None)
    if audio:
        meta["audio_codec"] = audio.get("codec_name")
        meta["channels"] = audio.get("channels")
        try:
            meta["sample_rate"] = int(audio.get("sample_rate"))
        except (TypeError, ValueError):
            pass
        try:
            meta["audio_bitrate"] = int(audio.get("bit_rate"))
        except (TypeError, ValueError):
            pass

    return meta


def resolution_text(meta):
    if meta and meta.get("width") and meta.get("height"):
        return f"{meta['width']}×{meta['height']}"
    return None


def generate_thumbnail(path, lut_path, output_path, at_seconds=1.0, width=320):
    """Extrage un cadru static (implicit ~1s in clip, sau `at_seconds` daca
    specificat — folosit de preview-ul interactiv, vezi media_preview.py)
    ca thumbnail, optional cu un LUT .cube aplicat prin filtrul nativ
    `lut3d` al FFmpeg — NU un player real-time. Formatul e dat de extensia
    din `output_path` — apelantul (main.py) foloseste `.png`, NU `.jpg`:
    `tk.PhotoImage` nativ (fara Pillow, dependinta pe care acest repo nu o
    are) suporta PNG dar nu JPEG. Mac (`MediaInspector.swift`) foloseste
    .jpg — NSImage citeste ambele formate nativ, nicio constrangere acolo.
    `width` (nou, 2026-09-05): 320 implicit pentru coada (rapid, mic);
    preview-ul fullscreen cere o latime mult mai mare (vezi media_preview.py)."""
    ffmpeg_path = get_ffmpeg_path()
    vf = f"scale={width}:-2"
    if lut_path:
        # ffmpeg cere backslash-escape pentru ':' din calea Windows
        # (C\:/...) in interiorul unui lant de filtre — vezi documentatia
        # oficiala a filtergraph-ului (caracterele : , [ ] au sens special).
        escaped = lut_path.replace("\\", "/").replace(":", "\\:")
        vf += f",lut3d=file='{escaped}'"
    try:
        result = subprocess.run(
            [ffmpeg_path, "-y", "-ss", f"{max(0.0, at_seconds):.3f}", "-i", path,
             "-frames:v", "1", "-vf", vf, output_path],
            capture_output=True, check=False, creationflags=_NO_WINDOW,
        )
        return result.returncode == 0 and os.path.isfile(output_path)
    except FileNotFoundError:
        return False


def thumbnails_folder():
    folder = os.path.join(tempfile.gettempdir(), "CGConvertorThumbs")
    os.makedirs(folder, exist_ok=True)
    return folder


def generate_html_report(jobs, statuses_text):
    """Raport HTML per lot — un singur fisier auto-continut (thumbnail-uri
    ca data URI base64). NEIMPLEMENTAT deliberat: varianta PDF — vezi
    CLAUDE.md pentru motiv."""
    rows = []
    for job in jobs:
        thumb_path = job.get("thumbnail_path")
        if thumb_path and os.path.isfile(thumb_path):
            with open(thumb_path, "rb") as f:
                b64 = base64.b64encode(f.read()).decode("ascii")
            mime = "image/png" if thumb_path.lower().endswith(".png") else "image/jpeg"
            thumb_tag = f'<img src="data:{mime};base64,{b64}" width="160">'
        else:
            thumb_tag = '<span class="muted">—</span>'

        meta = job.get("metadata") or {}
        meta_parts = [
            resolution_text(meta),
            (meta.get("video_codec") or "").upper() or None,
            f"{meta['frame_rate']} fps" if meta.get("frame_rate") else None,
            f"{meta['duration']:.1f}s" if meta.get("duration") else None,
            (meta.get("audio_codec") or "").upper() or None,
        ]
        meta_text = " · ".join(p for p in meta_parts if p) or "—"
        status_text = statuses_text.get(id(job), job.get("status", ""))

        rows.append(f"""
        <tr>
          <td>{thumb_tag}</td>
          <td>{os.path.basename(job["path"])}</td>
          <td>{meta_text}</td>
          <td>{status_text}</td>
        </tr>""")

    html = f"""<!DOCTYPE html><html><head><meta charset="utf-8">
<title>Raport CG Convertor</title>
<style>
body{{font-family:-apple-system,Helvetica,Arial,sans-serif;background:#14161A;color:#EDEFF2;padding:24px}}
h1{{color:#E8963C}}
table{{border-collapse:collapse;width:100%}}
td,th{{border-bottom:1px solid #2B2F36;padding:10px;text-align:left;vertical-align:middle}}
.muted{{color:#5C6169}}
</style></head><body>
<h1>Raport conversie — CG Convertor</h1>
<p>{len(jobs)} fișiere · generat {time.strftime("%Y-%m-%d %H:%M")}</p>
<table><tr><th>Thumbnail</th><th>Fișier</th><th>Metadata</th><th>Status</th></tr>
{"".join(rows)}
</table></body></html>"""

    output_path = os.path.join(tempfile.gettempdir(), f"CGConvertor_Raport_{int(time.time())}.html")
    with open(output_path, "w", encoding="utf-8") as f:
        f.write(html)
    return output_path
